# Writing Efficient Python Code

## Built-In Functions

In [9]:
# Create a range object that goes from 0 to 5
nums = range(6)
print(type(nums))

# Convert nums to a list
nums_list = list(nums)
print(nums_list)

# Create a new list of odd numbers from 1 to 11 by unpacking a range object
nums_list2 = [*range(1,12,2)]
print(nums_list2)

<class 'range'>
[0, 1, 2, 3, 4, 5]
[1, 3, 5, 7, 9, 11]


In [17]:
names = ['Jerry', 'Kramer', 'Elaine', 'George', 'Newman']

# Rewrite the for loop to use enumerate
indexed_names = []
for i,name in enumerate(names):
    index_name = (i,name)
    indexed_names.append(index_name) 
print(indexed_names)

# Rewrite the above for loop using list comprehension
indexed_names_comp = [(i,name) for i,name in enumerate(names)]
print(indexed_names_comp)

# Unpack an enumerate object with a starting index of one
indexed_names_unpack = [*enumerate(names, start=1)]
print(indexed_names_unpack)

[(0, 'Jerry'), (1, 'Kramer'), (2, 'Elaine'), (3, 'George'), (4, 'Newman')]
[(0, 'Jerry'), (1, 'Kramer'), (2, 'Elaine'), (3, 'George'), (4, 'Newman')]
[(1, 'Jerry'), (2, 'Kramer'), (3, 'Elaine'), (4, 'George'), (5, 'Newman')]


In [19]:
# Use map to apply str.upper to each element in names
names_map  = map(lambda x: x.upper(), names)

# Print the type of the names_map
print(type(names_map))

# Unpack names_map into a list
names_uppercase = [*names_map]

# Print the list created above
print(names_uppercase)

<class 'map'>
['JERRY', 'KRAMER', 'ELAINE', 'GEORGE', 'NEWMAN']


## Using Numpy Arrays

In [25]:
import numpy as np 

nums = np.array([[1,2,4,4,5],[6,7,9,9,10]])
print(nums)

# Print second row of nums
print(nums[1,:])

# Print all elements of nums that are greater than six
print(nums[nums > 6])

# Double every element of nums
nums_dbl = nums * 2
print(nums_dbl)

# Replace the third column of nums
nums[:,2] = nums[:,2] + 1
print(nums)

[[ 1  2  4  4  5]
 [ 6  7  9  9 10]]
[ 6  7  9  9 10]
[ 7  9  9 10]
[[ 2  4  8  8 10]
 [12 14 18 18 20]]
[[ 1  2  5  4  5]
 [ 6  7 10  9 10]]


In [31]:
# Create a list of arrival times
arrival_times = [*range(10,60,10)]
names = ['Jerry', 'Kramer', 'Elaine', 'George', 'Newman']
def welcome_guest(tup):
    name,time = tup
    return f"Welcome {name}. You are {time} minutes late."

# Convert arrival_times to an array and update the times
arrival_times_np = np.array(arrival_times)
new_times = arrival_times_np - 3

# Use list comprehension and enumerate to pair guests to new times
guest_arrivals = [(names[i],time) for i,time in enumerate(new_times)]

# Map the welcome_guest function to each (guest,time) pair
welcome_map = map(welcome_guest, guest_arrivals)

guest_welcomes = [*welcome_map]
print(*guest_welcomes, sep='\n')

Welcome Jerry. You are 7 minutes late.
Welcome Kramer. You are 17 minutes late.
Welcome Elaine. You are 27 minutes late.
Welcome George. You are 37 minutes late.
Welcome Newman. You are 47 minutes late.


## Examining Runtime

Calculate runtime with `%timeit`, which is a **magic command** like `pip`

In [32]:

#* apply timeit to the mapping
%timeit guest_welcomes = [*welcome_map]

133 ns ± 8.44 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


7 total runs each of 10 million loops, resulted in an average time of 133 nanoseconds with 8.44 nanoseconds of deviation.

In [36]:

#* specify number of loops
%timeit -r2 -n10 guest_welcomes = [*welcome_map] # 2 runs, 10 loops


239 ns ± 70.8 ns per loop (mean ± std. dev. of 2 runs, 10 loops each)


In [44]:

#* save output using -o
times = %timeit -o -r5 -n10 names_uppercase = list(map(lambda x: x.upper(), names))

#* See all runs
print(times.timings)
print(times.best)
print(times.worst)

2.11 μs ± 545 ns per loop (mean ± std. dev. of 5 runs, 10 loops each)
[2.3673004761803894e-06, 1.8359001842327417e-06, 3.05499997921288e-06, 1.5988996892701835e-06, 1.677100226515904e-06]
1.5988996892701835e-06
3.05499997921288e-06


## **Code Profiling for Runtime using `lineprofiler`**

In [2]:
%load_ext line_profiler
%lprun -f welcome_guest welcome_guest(('Kramer',5))

UsageError: Could not find module welcome_guest.
NameError: name 'welcome_guest' is not defined


- **Hits** - Number of time the line was executed
- **Time** - Total amount of time the line was executed
- **Time** - percent of total time the line was executed versus overall program

In [1]:
def convert_units(heroes, heights, weights):

    new_hts = [ht * 0.39370  for ht in heights]
    new_wts = [wt * 2.20462  for wt in weights]

    hero_data = {}

    for i,hero in enumerate(heroes):
        hero_data[hero] = (new_hts[i], new_wts[i])

    return hero_data
heroes = ['A-Bomb',
 'Abe Sapien',
 'Abin Sur',
 'Abomination',
 'Absorbing Man',
 'Adam Strange',
 'Agent 13',
 'Agent Bob',
 'Agent Zero',
 'Air-Walker',
 'Ajax',
 'Alan Scott',
 'Alfred Pennyworth',
 'Alien',
 'Amazo',
 'Ammo',
 'Angel',
 'Angel Dust',
 'Angel Salvadore',
 'Animal Man',
 'Annihilus',
 'Ant-Man',
 'Ant-Man II',
 'Anti-Venom',
 'Apocalypse',
 'Aqualad',
 'Aquaman',
 'Arachne',
 'Archangel',
 'Arclight',
 'Ardina',
 'Ares',
 'Ariel',
 'Armor',
 'Atlas',
 'Atom',
 'Atom Girl',
 'Atom II',
 'Aurora',
 'Azazel',
 'Bane',
 'Banshee',
 'Bantam',
 'Batgirl',
 'Batgirl IV',
 'Batgirl VI',
 'Batman',
 'Batman II',
 'Battlestar',
 'Beak',
 'Beast',
 'Beast Boy',
 'Beta Ray Bill',
 'Big Barda',
 'Big Man',
 'Binary',
 'Bishop',
 'Bizarro',
 'Black Adam',
 'Black Bolt',
 'Black Canary',
 'Black Cat',
 'Black Knight III',
 'Black Lightning',
 'Black Mamba',
 'Black Manta',
 'Black Panther',
 'Black Widow',
 'Black Widow II',
 'Blackout',
 'Blackwing',
 'Blackwulf',
 'Blade',
 'Bling!',
 'Blink',
 'Blizzard II',
 'Blob',
 'Bloodaxe',
 'Blue Beetle II',
 'Boom-Boom',
 'Booster Gold',
 'Box III',
 'Brainiac',
 'Brainiac 5',
 'Brother Voodoo',
 'Buffy',
 'Bullseye',
 'Bumblebee',
 'Cable',
 'Callisto',
 'Cannonball',
 'Captain America',
 'Captain Atom',
 'Captain Britain',
 'Captain Mar-vell',
 'Captain Marvel',
 'Captain Marvel II',
 'Carnage',
 'Cat',
 'Catwoman',
 'Cecilia Reyes',
 'Century',
 'Chamber',
 'Changeling',
 'Cheetah',
 'Cheetah II',
 'Cheetah III',
 'Chromos',
 'Citizen Steel',
 'Cloak',
 'Clock King',
 'Colossus',
 'Copycat',
 'Corsair',
 'Cottonmouth',
 'Crimson Dynamo',
 'Crystal',
 'Cyborg',
 'Cyclops',
 'Cypher',
 'Dagger',
 'Daredevil',
 'Darkhawk',
 'Darkseid',
 'Darkstar',
 'Darth Vader',
 'Dash',
 'Dazzler',
 'Deadman',
 'Deadpool',
 'Deadshot',
 'Deathlok',
 'Deathstroke',
 'Demogoblin',
 'Destroyer',
 'Diamondback',
 'Doc Samson',
 'Doctor Doom',
 'Doctor Doom II',
 'Doctor Fate',
 'Doctor Octopus',
 'Doctor Strange',
 'Domino',
 'Donna Troy',
 'Doomsday',
 'Doppelganger',
 'Drax the Destroyer',
 'Elastigirl',
 'Electro',
 'Elektra',
 'Elongated Man',
 'Emma Frost',
 'Enchantress',
 'Etrigan',
 'Evil Deadpool',
 'Evilhawk',
 'Exodus',
 'Fabian Cortez',
 'Falcon',
 'Feral',
 'Fin Fang Foom',
 'Firebird',
 'Firelord',
 'Firestar',
 'Firestorm',
 'Flash',
 'Flash II',
 'Flash III',
 'Flash IV',
 'Forge',
 'Franklin Richards',
 'Franklin Storm',
 'Frenzy',
 'Frigga',
 'Galactus',
 'Gambit',
 'Gamora',
 'Genesis',
 'Ghost Rider',
 'Giganta',
 'Gladiator',
 'Goblin Queen',
 'Goku',
 'Goliath IV',
 'Gorilla Grodd',
 'Granny Goodness',
 'Gravity',
 'Green Arrow',
 'Green Goblin',
 'Green Goblin II',
 'Green Goblin III',
 'Green Goblin IV',
 'Groot',
 'Guy Gardner',
 'Hal Jordan',
 'Han Solo',
 'Harley Quinn',
 'Havok',
 'Hawk',
 'Hawkeye',
 'Hawkeye II',
 'Hawkgirl',
 'Hawkman',
 'Hawkwoman',
 'Hawkwoman III',
 'Heat Wave',
 'Hela',
 'Hellboy',
 'Hellcat',
 'Hellstorm',
 'Hercules',
 'Hobgoblin',
 'Hope Summers',
 'Howard the Duck',
 'Hulk',
 'Human Torch',
 'Huntress',
 'Husk',
 'Hybrid',
 'Hydro-Man',
 'Hyperion',
 'Iceman',
 'Impulse',
 'Ink',
 'Invisible Woman',
 'Iron Fist',
 'Iron Man',
 'Jack of Hearts',
 'Jack-Jack',
 'James T. Kirk',
 'Jean Grey',
 'Jennifer Kale',
 'Jessica Jones',
 'Jigsaw',
 'John Stewart',
 'John Wraith',
 'Joker',
 'Jolt',
 'Jubilee',
 'Juggernaut',
 'Justice',
 'Kang',
 'Karate Kid',
 'Killer Croc',
 'Kilowog',
 'Kingpin',
 'Klaw',
 'Kraven II',
 'Kraven the Hunter',
 'Krypto',
 'Kyle Rayner',
 'Lady Deathstrike',
 'Leader',
 'Legion',
 'Lex Luthor',
 'Light Lass',
 'Lightning Lad',
 'Lightning Lord',
 'Living Brain',
 'Lizard',
 'Lobo',
 'Loki',
 'Longshot',
 'Luke Cage',
 'Luke Skywalker',
 'Mach-IV',
 'Machine Man',
 'Magneto',
 'Man-Thing',
 'Man-Wolf',
 'Mandarin',
 'Mantis',
 'Martian Manhunter',
 'Marvel Girl',
 'Master Brood',
 'Maverick',
 'Maxima',
 'Medusa',
 'Meltdown',
 'Mephisto',
 'Mera',
 'Metallo',
 'Metamorpho',
 'Metron',
 'Micro Lad',
 'Mimic',
 'Miss Martian',
 'Mister Fantastic',
 'Mister Freeze',
 'Mister Sinister',
 'Mockingbird',
 'MODOK',
 'Molten Man',
 'Monarch',
 'Moon Knight',
 'Moonstone',
 'Morlun',
 'Morph',
 'Moses Magnum',
 'Mr Immortal',
 'Mr Incredible',
 'Ms Marvel II',
 'Multiple Man',
 'Mysterio',
 'Mystique',
 'Namor',
 'Namora',
 'Namorita',
 'Naruto Uzumaki',
 'Nebula',
 'Nick Fury',
 'Nightcrawler',
 'Nightwing',
 'Northstar',
 'Nova',
 'Odin',
 'Omega Red',
 'Omniscient',
 'One Punch Man',
 'Onslaught',
 'Oracle',
 'Paul Blart',
 'Penance II',
 'Penguin',
 'Phantom Girl',
 'Phoenix',
 'Plantman',
 'Plastic Man',
 'Plastique',
 'Poison Ivy',
 'Polaris',
 'Power Girl',
 'Predator',
 'Professor X',
 'Professor Zoom',
 'Psylocke',
 'Punisher',
 'Purple Man',
 'Pyro',
 'Question',
 'Quicksilver',
 'Quill',
 "Ra's Al Ghul",
 'Raven',
 'Ray',
 'Razor-Fist II',
 'Red Arrow',
 'Red Hood',
 'Red Hulk',
 'Red Robin',
 'Red Skull',
 'Red Tornado',
 'Rhino',
 'Rick Flag',
 'Ripcord',
 'Robin',
 'Robin II',
 'Robin III',
 'Robin V',
 'Rocket Raccoon',
 'Rogue',
 'Ronin',
 'Rorschach',
 'Sabretooth',
 'Sage',
 'Sandman',
 'Sasquatch',
 'Scarecrow',
 'Scarlet Spider',
 'Scarlet Spider II',
 'Scarlet Witch',
 'Scorpion',
 'Sentry',
 'Shadow King',
 'Shadow Lass',
 'Shadowcat',
 'Shang-Chi',
 'Shatterstar',
 'She-Hulk',
 'She-Thing',
 'Shocker',
 'Shriek',
 'Sif',
 'Silver Surfer',
 'Silverclaw',
 'Sinestro',
 'Siren',
 'Siryn',
 'Skaar',
 'Snowbird',
 'Solomon Grundy',
 'Songbird',
 'Space Ghost',
 'Spawn',
 'Spider-Girl',
 'Spider-Gwen',
 'Spider-Man',
 'Spider-Woman',
 'Spider-Woman III',
 'Spider-Woman IV',
 'Spock',
 'Spyke',
 'Star-Lord',
 'Starfire',
 'Stargirl',
 'Static',
 'Steel',
 'Steppenwolf',
 'Storm',
 'Sunspot',
 'Superboy',
 'Superboy-Prime',
 'Supergirl',
 'Superman',
 'Swarm',
 'Synch',
 'T-1000',
 'Taskmaster',
 'Tempest',
 'Thanos',
 'The Comedian',
 'Thing',
 'Thor',
 'Thor Girl',
 'Thunderbird',
 'Thunderbird III',
 'Thunderstrike',
 'Thundra',
 'Tiger Shark',
 'Tigra',
 'Tinkerer',
 'Toad',
 'Toxin',
 'Trickster',
 'Triplicate Girl',
 'Triton',
 'Two-Face',
 'Ultragirl',
 'Ultron',
 'Utgard-Loki',
 'Vagabond',
 'Valerie Hart',
 'Valkyrie',
 'Vanisher',
 'Vegeta',
 'Venom',
 'Venom II',
 'Venom III',
 'Vertigo II',
 'Vibe',
 'Vindicator',
 'Violet Parr',
 'Vision',
 'Vision II',
 'Vixen',
 'Vulture',
 'Walrus',
 'War Machine',
 'Warbird',
 'Warlock',
 'Warp',
 'Warpath',
 'Wasp',
 'White Queen',
 'Winter Soldier',
 'Wiz Kid',
 'Wolfsbane',
 'Wolverine',
 'Wonder Girl',
 'Wonder Man',
 'Wonder Woman',
 'Wyatt Wingfoot',
 'X-23',
 'X-Man',
 'Yellow Claw',
 'Yellowjacket',
 'Yellowjacket II',
 'Yoda',
 'Zatanna',
 'Zoom']

import numpy as np
hts = np.array([203. , 191. , 185. , 203. , 193. , 185. , 173. , 178. , 191. ,
       188. , 193. , 180. , 178. , 244. , 257. , 188. , 183. , 165. ,
       163. , 183. , 180. , 211. , 183. , 229. , 213. , 178. , 185. ,
       175. , 183. , 173. , 193. , 185. , 165. , 163. , 183. , 178. ,
       168. , 183. , 180. , 183. , 203. , 183. , 165. , 170. , 165. ,
       168. , 188. , 178. , 198. , 175. , 180. , 173. , 201. , 188. ,
       165. , 180. , 198. , 191. , 191. , 188. , 165. , 178. , 183. ,
       185. , 170. , 188. , 183. , 170. , 170. , 191. , 185. , 188. ,
       188. , 168. , 165. , 175. , 178. , 218. , 183. , 165. , 196. ,
       193. , 198. , 170. , 183. , 157. , 183. , 170. , 203. , 175. ,
       183. , 188. , 193. , 198. , 188. , 180. , 175. , 185. , 173. ,
       175. , 170. , 201. , 175. , 180. , 163. , 170. , 175. , 185. ,
       183. , 226. , 178. , 226. , 183. , 191. , 183. , 180. , 168. ,
       198. , 191. , 175. , 165. , 183. , 185. , 267. , 168. , 198. ,
       122. , 173. , 183. , 188. , 185. , 193. , 193. , 185. , 188. ,
       193. , 198. , 201. , 201. , 188. , 175. , 188. , 173. , 175. ,
       244. , 196. , 193. , 168. , 180. , 175. , 185. , 178. , 168. ,
       193. , 188. , 191. , 183. , 196. , 188. , 175. , 975. , 165. ,
       193. , 173. , 188. , 180. , 183. , 183. , 157. , 183. , 142. ,
       188. , 211. , 180. , 876. , 185. , 183. , 185. , 188. ,  62.5,
       198. , 168. , 175. , 183. , 198. , 178. , 178. , 188. , 180. ,
       178. , 183. , 178. , 701. , 188. , 188. , 183. , 170. , 183. ,
       185. , 191. , 165. , 175. , 185. , 175. , 170. , 180. , 213. ,
       259. , 173. , 185. , 196. , 180. , 168. ,  79. , 244. , 178. ,
       180. , 170. , 175. , 188. , 183. , 173. , 170. , 180. , 168. ,
       180. , 198. , 155. ,  71. , 178. , 168. , 168. , 170. , 188. ,
       185. , 183. , 196. , 165. , 165. , 287. , 178. , 191. , 173. ,
       244. , 234. , 201. , 188. , 191. , 183. ,  64. , 180. , 175. ,
       178. , 175. , 188. , 165. , 155. , 191. , 198. , 203. , 229. ,
       193. , 188. , 198. , 168. , 180. , 183. , 188. , 213. , 188. ,
       188. , 168. , 201. , 170. , 183. , 193. , 180. , 180. , 165. ,
       198. , 175. , 196. , 185. , 185. , 183. , 188. , 178. , 185. ,
       183. , 196. , 175. , 366. , 196. , 193. , 188. , 180. , 188. ,
       178. , 175. , 188. , 201. , 173. , 180. , 180. , 178. , 188. ,
       180. , 168. , 168. , 185. , 185. , 175. , 178. , 180. , 185. ,
       206. , 211. , 180. , 175. , 305. , 178. , 170. , 183. , 157. ,
       168. , 168. , 183. , 185. , 168. , 168. , 170. , 180. , 213. ,
       183. , 180. , 180. , 183. , 180. , 178. , 188. , 183. , 163. ,
       193. , 165. , 178. , 191. , 180. , 183. , 213. , 165. , 188. ,
       185. , 196. , 185. , 180. , 178. , 183. , 165. , 137. , 122. ,
       173. , 191. , 168. , 198. , 170. , 185. , 305. , 183. , 178. ,
       193. , 170. , 211. , 188. , 185. , 173. , 168. , 178. , 191. ,
       201. , 183. , 175. , 173. , 188. , 193. , 157. , 201. , 175. ,
       168. , 198. , 178. , 279. , 165. , 188. , 211. , 170. , 165. ,
       178. , 178. , 173. , 178. , 185. , 183. , 188. , 193. , 165. ,
       170. , 201. , 183. , 180. , 173. , 170. , 180. , 165. , 191. ,
       196. , 180. , 183. , 188. , 163. , 201. , 188. , 183. , 198. ,
       175. , 185. , 175. , 198. , 218. , 185. , 178. , 163. , 175. ,
       188. , 183. , 168. , 188. , 183. , 168. , 206. ,  15.2, 168. ,
       175. , 191. , 165. , 168. , 191. , 175. , 229. , 168. , 178. ,
       165. , 137. , 191. , 191. , 175. , 180. , 183. , 185. , 180. ,
       188. , 173. , 218. , 163. , 178. , 175. , 140. , 366. , 160. ,
       165. , 188. , 183. , 196. , 155. , 175. , 188. , 183. , 165. ,
        66. , 170. , 185. ])
wts = np.array([441.,  65.,  90., 441., 122.,  88.,  61.,  81., 104., 108.,  90.,
        90.,  72., 169., 173., 101.,  68.,  57.,  54.,  83.,  90., 122.,
        86., 358., 135., 106., 146.,  63.,  68.,  57.,  98., 270.,  59.,
        50., 101.,  68.,  54.,  81.,  63.,  67., 180.,  77.,  54.,  57.,
        52.,  61.,  95.,  79., 133.,  63., 181.,  68., 216., 135.,  71.,
        54., 124., 155., 113.,  95.,  58.,  54.,  86.,  90.,  52.,  92.,
        90.,  59.,  61., 104.,  86.,  88.,  97.,  68.,  56.,  77., 230.,
       495.,  86.,  55.,  97., 110., 135.,  61.,  99.,  52.,  90.,  59.,
       158.,  74.,  81., 108.,  90., 116., 108.,  74.,  74.,  86.,  61.,
        61.,  62.,  97.,  63.,  81.,  50.,  55.,  54.,  86., 170.,  70.,
        78., 225.,  67.,  79.,  99., 104.,  50., 173.,  88.,  68.,  52.,
        90.,  81., 817.,  56., 135.,  27.,  52.,  90.,  95.,  91., 178.,
       101.,  95., 383.,  90., 171., 187., 132.,  89., 110.,  81.,  54.,
        63., 412., 104., 306.,  56.,  74.,  59.,  80.,  65.,  57., 203.,
        95., 106.,  88.,  96., 108.,  50.,  18.,  56.,  99.,  56.,  91.,
        81.,  88.,  86.,  52.,  81.,  45.,  92., 104., 167.,  16.,  81.,
        77.,  86.,  99., 630., 268.,  50.,  62.,  90., 270., 115.,  79.,
        88.,  83.,  77.,  88.,  79.,   4.,  95.,  90.,  79.,  63.,  79.,
        89., 104.,  57.,  61.,  88.,  54.,  65.,  81., 225., 158.,  61.,
        81., 146.,  83.,  48.,  18., 630.,  77.,  59.,  58.,  77., 119.,
       207.,  65.,  65.,  81.,  54.,  79., 191.,  79.,  14.,  77.,  52.,
        55.,  56., 113.,  90.,  88.,  86.,  49.,  52., 855.,  81., 104.,
        72., 356., 324., 203.,  97.,  99., 106.,  18.,  79.,  58.,  63.,
        59.,  95.,  54.,  65.,  95., 360., 230., 288., 236.,  36., 191.,
        77.,  79., 383.,  86., 225.,  90.,  97.,  52., 135.,  56.,  81.,
       110.,  72.,  59.,  54., 140.,  72.,  90.,  90.,  86.,  77., 101.,
        61.,  81.,  86., 128.,  61., 338., 248.,  90., 101.,  59.,  79.,
        79.,  72.,  70., 158.,  61.,  70.,  79.,  54., 125.,  85., 101.,
        54.,  83.,  99.,  88.,  79.,  83.,  86., 293., 191.,  65.,  69.,
       405.,  59., 117.,  89.,  79.,  54.,  52.,  87.,  80.,  55.,  50.,
        52.,  81., 234.,  86.,  81.,  70.,  90.,  74.,  68.,  83.,  79.,
        56.,  97.,  50.,  70., 117.,  83.,  81., 630.,  56., 108., 146.,
       320.,  85.,  72.,  79., 101.,  56.,  38.,  25.,  54., 104.,  63.,
       171.,  61., 203., 900.,  63.,  74., 113.,  59., 310.,  87., 149.,
        54.,  50.,  79.,  88., 315., 153.,  79.,  52., 191., 101.,  50.,
        92.,  72.,  52., 180.,  49., 437.,  65., 113., 405.,  54.,  56.,
        74.,  59.,  55.,  58.,  81.,  83.,  79.,  71.,  62.,  63., 131.,
        91.,  57.,  77.,  68.,  77.,  54., 101.,  47.,  74., 146.,  99.,
        54., 443., 101., 225., 288., 143., 101.,  74., 288., 158., 203.,
        81.,  54.,  76.,  97.,  81.,  59.,  86.,  82., 105., 331.,  58.,
        54.,  56., 214.,  79.,  73., 117.,  50., 334.,  52.,  71.,  54.,
        41., 135., 135.,  63.,  79., 162.,  95.,  54., 108.,  67., 158.,
        50.,  65., 117.,  39., 473., 135.,  51., 171.,  74., 117.,  50.,
        61.,  95.,  83.,  52.,  17.,  57.,  81.])

%lprun -f convert_units convert_units(heroes,hts,wts)

def convert_units_broadcast(heroes, heights, weights):

    # Array broadcasting instead of list comprehension
    new_hts = heights * 0.39370
    new_wts = weights * 2.20462

    hero_data = {}

    for i,hero in enumerate(heroes):
        hero_data[hero] = (new_hts[i], new_wts[i])

    return hero_data

%lprun -f convert_units_broadcast convert_units_broadcast(heroes,hts,wts)

UsageError: Line magic function `%lprun` not found.


## Code Profiling for Memory Usage

- Functions must be importhed when using `memory_profiler`

In [54]:
from funcs import convert_units, convert_units_broadcast

%load_ext memory_profiler
%mprun -f convert_units convert_units(heroes,hts,wts)


Filename: /home/toni11/Documents/Programming_Projects/Data_Science_Projects/Data_Engineering/funcs.py

Line #    Mem usage    Increment  Occurrences   Line Contents
     1     95.5 MiB     95.5 MiB           1   def convert_units(heroes, heights, weights):
     2                                         
     3     95.7 MiB      0.1 MiB         481       new_hts = [ht * 0.39370  for ht in heights]
     4     95.7 MiB      0.0 MiB         481       new_wts = [wt * 2.20462  for wt in weights]
     5                                         
     6     95.7 MiB      0.0 MiB           1       hero_data = {}
     7                                         
     8     95.7 MiB      0.0 MiB         481       for i,hero in enumerate(heroes):
     9     95.7 MiB      0.0 MiB         480           hero_data[hero] = (new_hts[i], new_wts[i])
    10                                         
    11     95.7 MiB      0.0 MiB           1       return hero_data

- **Increment** - Memory usage of current line versus previous line

In [56]:
%mprun -f convert_units_broadcast convert_units_broadcast(heroes,hts,wts)

Filename: /home/toni11/Documents/Programming_Projects/Data_Science_Projects/Data_Engineering/funcs.py

Line #    Mem usage    Increment  Occurrences   Line Contents
    12     95.7 MiB     95.7 MiB           1   def convert_units_broadcast(heroes, heights, weights):
    13                                         
    14                                             # Array broadcasting instead of list comprehension
    15     95.7 MiB      0.0 MiB           1       new_hts = heights * 0.39370
    16     95.7 MiB      0.0 MiB           1       new_wts = weights * 2.20462
    17                                         
    18     95.7 MiB      0.0 MiB           1       hero_data = {}
    19                                         
    20     95.7 MiB      0.0 MiB         481       for i,hero in enumerate(heroes):
    21     95.7 MiB      0.0 MiB         480           hero_data[hero] = (new_hts[i], new_wts[i])
    22                                         
    23     95.7 MiB      0.0 MiB 

## `itertools` and `Collections`

In [14]:
heroes =['A-Bomb',
 'Abe Sapien',
 'Agent Bob',
 'Ant-Man II',
 'Anti-Venom',
 'Apocalypse',
 'Aqualad',
 'Aquaman',
 'Arachne',
 'Archangel',
 'Arclight',
 'Banshee',
 'Bantam',
 'Batgirl',
 'Batgirl IV',
 'Batgirl VI',
 'Batman',
 'Batman II',
 'Battlestar',
 'Beak',
 'Beast',
 'Beast Boy',
 'Beta Ray Bill',
 'Big Barda',
 'Big Man',
 'Binary',
 'Bishop',
 'Bizarro',
 'Black Adam',
 'Black Bolt',
 'Bling!',
 'Blink',
 'Blizzard II',
 'Blob',
 'Bloodaxe',
 'Blue Beetle II',
 'Boom-Boom',
 'Booster Gold',
 'Box III',
 'Brainiac',
 'Brainiac 5',
 'Brother Voodoo',
 'Buffy',
 'Bullseye',
 'Bumblebee',
 'Cable',
 'Callisto',
 'Cannonball',
 'Captain America',
 'Captain Atom',
 'Captain Britain',
 'Captain Mar-vell',
 'Captain Marvel',
 'Captain Marvel II',
 'Carnage',
 'Cat',
 'Catwoman',
 'Cecilia Reyes',
 'Century',
 'Chamber',
 'Changeling',
 'Cheetah',
 'Cheetah II',
 'Cheetah III',
 'Chromos',
 'Citizen Steel',
 'Cloak',
 'Clock King',
 'Deadpool',
 'Deadshot',
 'Deathlok',
 'Deathstroke',
 'Demogoblin',
 'Destroyer',
 'Diamondback',
 'Doc Samson',
 'Doctor Doom',
 'Doctor Doom II',
 'Doctor Fate',
 'Doppelganger',
 'Drax the Destroyer',
 'Elastigirl',
 'Electro',
 'Elektra',
 'Exodus',
 'Fabian Cortez',
 'Falcon',
 'Feral',
 'Fin Fang Foom',
 'Firebird',
 'Firelord',
 'Frigga',
 'Galactus',
 'Gambit',
 'Gamora',
 'Genesis',
 'Green Goblin',
 'Green Goblin II',
 'Green Goblin III',
 'Green Goblin IV',
 'Groot',
 'Guy Gardner',
 'Hal Jordan',
 'Han Solo',
 'Harley Quinn',
 'Havok',
 'Hawk',
 'Hawkeye',
 'Hawkeye II',
 'Hawkgirl',
 'Hawkman',
 'Hawkwoman',
 'Hawkwoman III',
 'Heat Wave',
 'Hela',
 'Hellboy',
 'Hybrid',
 'Hydro-Man',
 'Hyperion',
 'Iceman',
 'Impulse',
 'Ink',
 'Invisible Woman',
 'Iron Fist',
 'Jolt',
 'Jubilee',
 'Juggernaut',
 'Justice',
 'Kang',
 'Karate Kid',
 'Killer Croc',
 'Kilowog',
 'Legion',
 'Lex Luthor',
 'Light Lass',
 'Lightning Lad',
 'Lightning Lord',
 'Living Brain',
 'Lizard',
 'Lobo',
 'Loki',
 'Longshot',
 'Luke Cage',
 'Luke Skywalker',
 'Mach-IV',
 'Machine Man',
 'Metallo',
 'Metamorpho',
 'Metron',
 'Micro Lad',
 'Monarch',
 'Moon Knight',
 'Moonstone',
 'Morlun',
 'Namora',
 'Namorita',
 'Naruto Uzumaki',
 'Oracle',
 'Paul Blart',
 'Penance II',
 'Penguin',
 'Phantom Girl',
 'Phoenix',
 'Plantman',
 'Plastic Man',
 'Red Arrow',
 'Red Hood',
 'Red Hulk',
 'Red Robin',
 'Red Skull',
 'Rogue',
 'Ronin',
 'Rorschach',
 'Sabretooth',
 'Sage',
 'Sandman',
 'Sasquatch',
 'Venom',
 'Venom II',
 'Venom III',
 'Vertigo II',
 'Vibe',
 'Vindicator',
 'Violet Parr',
 'Vision',
 'Vision II',
 'Vixen',
 'Vulture',
 'Wyatt Wingfoot',
 'X-23',
 'X-Man',
 'Zoom']

from collections import Counter

hero_count = Counter(heroes)
print(hero_count, '\n')

Counter({'A-Bomb': 1, 'Abe Sapien': 1, 'Agent Bob': 1, 'Ant-Man II': 1, 'Anti-Venom': 1, 'Apocalypse': 1, 'Aqualad': 1, 'Aquaman': 1, 'Arachne': 1, 'Archangel': 1, 'Arclight': 1, 'Banshee': 1, 'Bantam': 1, 'Batgirl': 1, 'Batgirl IV': 1, 'Batgirl VI': 1, 'Batman': 1, 'Batman II': 1, 'Battlestar': 1, 'Beak': 1, 'Beast': 1, 'Beast Boy': 1, 'Beta Ray Bill': 1, 'Big Barda': 1, 'Big Man': 1, 'Binary': 1, 'Bishop': 1, 'Bizarro': 1, 'Black Adam': 1, 'Black Bolt': 1, 'Bling!': 1, 'Blink': 1, 'Blizzard II': 1, 'Blob': 1, 'Bloodaxe': 1, 'Blue Beetle II': 1, 'Boom-Boom': 1, 'Booster Gold': 1, 'Box III': 1, 'Brainiac': 1, 'Brainiac 5': 1, 'Brother Voodoo': 1, 'Buffy': 1, 'Bullseye': 1, 'Bumblebee': 1, 'Cable': 1, 'Callisto': 1, 'Cannonball': 1, 'Captain America': 1, 'Captain Atom': 1, 'Captain Britain': 1, 'Captain Mar-vell': 1, 'Captain Marvel': 1, 'Captain Marvel II': 1, 'Carnage': 1, 'Cat': 1, 'Catwoman': 1, 'Cecilia Reyes': 1, 'Century': 1, 'Chamber': 1, 'Changeling': 1, 'Cheetah': 1, 'Cheetah 

In [ ]:
# Import combinations from itertools
from itertools import combinations

# Collect all possible combinations of 4 Pokémon directly into a list
combos_4 = [*combinations(heroes,4)]
print(combos_4)

## Using Set

- Membership testing is faster with set than list. However, comverting a list to a set will add space complexity.

In [6]:
ash_pokedex = ['Pikachu', 'Bulbasaur', 'Koffing', 'Spearow', 'Vulpix', 'Wigglytuff', 'Zubat', 'Rattata', 'Psyduck', 'Squirtle'] 
misty_pokedex = ['Krabby', 'Horsea', 'Slowbro', 'Tentacool', 'Vaporeon', 'Magikarp', 'Poliwag', 'Starmie', 'Psyduck', 'Squirtle']
misty_pokedex_set = set(misty_pokedex)

%timeit -r1 -n10000 ash_check = "Pikachu" in ash_pokedex
%timeit -r1 -n10000 misty_check = "Pikachu" in misty_pokedex_set

#* However, adding the conversion of a list to a set will add additional time
%timeit -r1 -n10000 misty_check = "Pikachu" in set(misty_pokedex)

95.7 ns ± 0 ns per loop (mean ± std. dev. of 1 run, 10,000 loops each)
84.9 ns ± 0 ns per loop (mean ± std. dev. of 1 run, 10,000 loops each)
291 ns ± 0 ns per loop (mean ± std. dev. of 1 run, 10,000 loops each)


In [10]:
from collections import Counter

#* Writing a better loop
generations = [4,
 1,
 3,
 5,
 1,
 3,
 2,
 1,
 5,
 3,
 6,
 4,
 5,
 2,
 3,
 1,
 1,
 4,
 5,
 5,
 2,
 3,
 6,
 3,
 1,
 5,
 6,
 6,
 5,
 4,
 2,
 3,
 3,
 3,
 3,
 6,
 3,
 5,
 4,
 2,
 5,
 3,
 1,
 5,
 3,
 2,
 1,
 6,
 4,
 4,
 6,
 5,
 1,
 3,
 2,
 5,
 5,
 4,
 5,
 6,
 5,
 3,
 4,
 4,
 4,
 4,
 1,
 4,
 6,
 4,
 1,
 3,
 3,
 3,
 6,
 4,
 5,
 3,
 3,
 3,
 1,
 2,
 5,
 1,
 1,
 1,
 1,
 4,
 4,
 4,
 6,
 6,
 2,
 4,
 3,
 2,
 4,
 5,
 3,
 6,
 6,
 3,
 1,
 1,
 2,
 1,
 5,
 5,
 4,
 3,
 5,
 3,
 2,
 5,
 3,
 4,
 3,
 4,
 4,
 2,
 2,
 5,
 5,
 5,
 1,
 2,
 4,
 5,
 5,
 5,
 6,
 5,
 5,
 3,
 2,
 6,
 1,
 5,
 4,
 6,
 6,
 1,
 1,
 1,
 1,
 2,
 6,
 6,
 1,
 1,
 4,
 1,
 4,
 4,
 5,
 1,
 5,
 5,
 1,
 2,
 5,
 5,
 3,
 4,
 3,
 3,
 5,
 5,
 5,
 1,
 1,
 1,
 4,
 3,
 1,
 2,
 5,
 5,
 5,
 4,
 2,
 5,
 2,
 6,
 5,
 1,
 1,
 3,
 1,
 1,
 3,
 6,
 2,
 5,
 5,
 4,
 2,
 6,
 1,
 6,
 6,
 4,
 6,
 6,
 3,
 5,
 2,
 5,
 5,
 6,
 6,
 4,
 6,
 2,
 4,
 4,
 5,
 5,
 4,
 3,
 1,
 4,
 5,
 1,
 1,
 4,
 5,
 2,
 4,
 3,
 4,
 2,
 4,
 1,
 6,
 1,
 1,
 1,
 1,
 5,
 5,
 6,
 6,
 3,
 5,
 5,
 5,
 2,
 1,
 6,
 1,
 4,
 3,
 3,
 3,
 1,
 3,
 3,
 5,
 1,
 4,
 3,
 1,
 6,
 5,
 5,
 4,
 6,
 6,
 2,
 5,
 4,
 4,
 1,
 1,
 2,
 2,
 4,
 6,
 2,
 2,
 1,
 2,
 2,
 3,
 5,
 1,
 2,
 3,
 4,
 6,
 1,
 5,
 1,
 3,
 1,
 5,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 5,
 3,
 2,
 1,
 3,
 5,
 6,
 5,
 5,
 1,
 1,
 4,
 4,
 5,
 5,
 3,
 3,
 5,
 5,
 5,
 3,
 5,
 2,
 1,
 5,
 2,
 3,
 3,
 4,
 5,
 2,
 2,
 4,
 1,
 5,
 3,
 5,
 5,
 3,
 6,
 5,
 3,
 4,
 3,
 3,
 4,
 3,
 2,
 4,
 3,
 3,
 4,
 4,
 1,
 1,
 1,
 2,
 2,
 1,
 1,
 4,
 1,
 1,
 4,
 3,
 6,
 4,
 4,
 5,
 3,
 1,
 2,
 4,
 5,
 2,
 2,
 1,
 3,
 3,
 3,
 3,
 3,
 6,
 6,
 1,
 4,
 3,
 3,
 1,
 1,
 1,
 5,
 5,
 3,
 3,
 2,
 4,
 5,
 3,
 2,
 4,
 1,
 4,
 4,
 1,
 3,
 1,
 4,
 5,
 2,
 5,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 3,
 1,
 3,
 2,
 6,
 6,
 3,
 3,
 3,
 2,
 1,
 1,
 1,
 1,
 5,
 4,
 4,
 5,
 6,
 6,
 5,
 5,
 5,
 1,
 1,
 5,
 5,
 3,
 1,
 5,
 2,
 6,
 4,
 2,
 1,
 1,
 1,
 5,
 5,
 1,
 2,
 2,
 1,
 4,
 3,
 2,
 1,
 1,
 1,
 1,
 3,
 1,
 4,
 2,
 1,
 4,
 4,
 1,
 2,
 5,
 4,
 6,
 2,
 2,
 6,
 2,
 1,
 2,
 3,
 4,
 1,
 1,
 1,
 3,
 3,
 4,
 3,
 3,
 3,
 2,
 5,
 5,
 1,
 1,
 4,
 4,
 5,
 3,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 5,
 3,
 3,
 5,
 5,
 1,
 1,
 5,
 5,
 6,
 3,
 2,
 5,
 5,
 5,
 1,
 1,
 1,
 3,
 3,
 1,
 5,
 2,
 5,
 5,
 3,
 5,
 3,
 3,
 3,
 1,
 4,
 5,
 4,
 3,
 4,
 3,
 2,
 3,
 5,
 3,
 5,
 5,
 5,
 2,
 6,
 2,
 3,
 4,
 6,
 4,
 3,
 3,
 6,
 1,
 2,
 1,
 2,
 6,
 2,
 2,
 2,
 5,
 1,
 3,
 4,
 2,
 5,
 3,
 1,
 6,
 3,
 2,
 3,
 4,
 3,
 6,
 1,
 2,
 4,
 4,
 4,
 1,
 1,
 2,
 5,
 5,
 4,
 2,
 2,
 2,
 2,
 3,
 3,
 5,
 3,
 3,
 5,
 3,
 2,
 6,
 5,
 6,
 3,
 6,
 1,
 4,
 1,
 2,
 1,
 1,
 5,
 5,
 5,
 5,
 5,
 4,
 2,
 2,
 3,
 3,
 4,
 2,
 4,
 5,
 3,
 3,
 6,
 3,
 5,
 4,
 5,
 5,
 2,
 2,
 6,
 2,
 6,
 2,
 5,
 2,
 2,
 4,
 5,
 5,
 5,
 1,
 5,
 1,
 1,
 1,
 4,
 3,
 5,
 1,
 3,
 1,
 5,
 6,
 3,
 6,
 5,
 1,
 5,
 1,
 3,
 3,
 3,
 1,
 5,
 4,
 1,
 1,
 1,
 5,
 5,
 3,
 3,
 1,
 3,
 2,
 5,
 2,
 4,
 4,
 4,
 3,
 3,
 2,
 6,
 5,
 2,
 4,
 6,
 3,
 1,
 5,
 5,
 3,
 5,
 5,
 1,
 5]

# Collect the count of each generation
gen_counts = Counter(generations)

# Improve for loop by moving one calculation above the loop
total_count = len(generations)

for gen,count in gen_counts.items():
    gen_percent = round(count / total_count * 100, 2)
    print('generation {}: count = {:3} percentage = {}'
          .format(gen, count, gen_percent))

generation 4: count = 112 percentage = 15.56
generation 1: count = 151 percentage = 20.97
generation 3: count = 136 percentage = 18.89
generation 5: count = 154 percentage = 21.39
generation 2: count =  99 percentage = 13.75
generation 6: count =  68 percentage = 9.44


In [15]:
from itertools import combinations
# Collect all possible pairs using combinations()
possible_pairs = [*combinations(heroes, 2)]

# Create an empty list called enumerated_tuples
enumerated_tuples = []

for i,pair in enumerate(possible_pairs, 1):
    enumerated_pair_tuple = (i,) + pair
    enumerated_tuples.append(enumerated_pair_tuple)

# Convert all tuples in enumerated_tuples to a list
enumerated_pairs = [*map(list, enumerated_tuples)]
print(enumerated_pairs)

[[1, 'A-Bomb', 'Abe Sapien'], [2, 'A-Bomb', 'Agent Bob'], [3, 'A-Bomb', 'Ant-Man II'], [4, 'A-Bomb', 'Anti-Venom'], [5, 'A-Bomb', 'Apocalypse'], [6, 'A-Bomb', 'Aqualad'], [7, 'A-Bomb', 'Aquaman'], [8, 'A-Bomb', 'Arachne'], [9, 'A-Bomb', 'Archangel'], [10, 'A-Bomb', 'Arclight'], [11, 'A-Bomb', 'Banshee'], [12, 'A-Bomb', 'Bantam'], [13, 'A-Bomb', 'Batgirl'], [14, 'A-Bomb', 'Batgirl IV'], [15, 'A-Bomb', 'Batgirl VI'], [16, 'A-Bomb', 'Batman'], [17, 'A-Bomb', 'Batman II'], [18, 'A-Bomb', 'Battlestar'], [19, 'A-Bomb', 'Beak'], [20, 'A-Bomb', 'Beast'], [21, 'A-Bomb', 'Beast Boy'], [22, 'A-Bomb', 'Beta Ray Bill'], [23, 'A-Bomb', 'Big Barda'], [24, 'A-Bomb', 'Big Man'], [25, 'A-Bomb', 'Binary'], [26, 'A-Bomb', 'Bishop'], [27, 'A-Bomb', 'Bizarro'], [28, 'A-Bomb', 'Black Adam'], [29, 'A-Bomb', 'Black Bolt'], [30, 'A-Bomb', 'Bling!'], [31, 'A-Bomb', 'Blink'], [32, 'A-Bomb', 'Blizzard II'], [33, 'A-Bomb', 'Blob'], [34, 'A-Bomb', 'Bloodaxe'], [35, 'A-Bomb', 'Blue Beetle II'], [36, 'A-Bomb', 'Boom-B

## Pandas DataFrame Iteration

In [17]:
import pandas as pd

# Define the data
data = {
    'Team': ['SFG', 'SFG', 'SFG', 'SFG', 'SFG'],
    'League': ['NL', 'NL', 'NL', 'NL', 'NL'],
    'Year': [2012, 2011, 2010, 2009, 2008],
    'RS': [718, 570, 697, 657, 640],
    'RA': [649, 578, 583, 611, 759],
    'W': [94, 86, 92, 88, 72],
    'G': [162, 162, 162, 162, 162],
    'Playoffs': [1, 0, 1, 0, 0]
}

# Create the DataFrame
giants_df = pd.DataFrame(data)

def calc_run_diff(runs_scored, runs_allowed):

    run_diff = runs_scored - runs_allowed

    return run_diff
#* Using iterrows() is more efficient than iloc

# Create an empty list to store run differentials
run_diffs = []

# Write a for loop and collect runs allowed and runs scored for each row
for i,row in giants_df.iterrows():
    runs_scored = row['RS']
    runs_allowed = row['RA']
    
    # Use the provided function to calculate run_diff for each row
    run_diff = calc_run_diff(runs_scored, runs_allowed)
    
    # Append each run differential to the output list
    run_diffs.append(run_diff)

giants_df['RD'] = run_diffs
display(giants_df.head())

Team League  Year   RS   RA   W    G  Playoffs   RD
0  SFG     NL  2012  718  649  94  162         1   69
1  SFG     NL  2011  570  578  86  162         0   -8
2  SFG     NL  2010  697  583  92  162         1  114
3  SFG     NL  2009  657  611  88  162         0   46
4  SFG     NL  2008  640  759  72  162         0 -119

## Third Best Method: `apply`

In [19]:
def calc_run_diff_apply(df):
    runs_scored = df['RS']
    runs_allowed = df['RA']
    run_diff = runs_scored - runs_allowed

    return run_diff\

giants_df['RD_apply'] = giants_df.apply(calc_run_diff_apply,axis=1)
display(giants_df.head())

Team League  Year   RS   RA   W    G  Playoffs   RD  RD_apply
0  SFG     NL  2012  718  649  94  162         1   69        69
1  SFG     NL  2011  570  578  86  162         0   -8        -8
2  SFG     NL  2010  697  583  92  162         1  114       114
3  SFG     NL  2009  657  611  88  162         0   46        46
4  SFG     NL  2008  640  759  72  162         0 -119      -119

In [22]:
# Display the first five rows of the DataFrame
print(giants_df.head())

# Create a win percentage Series 
win_percs = giants_df.apply(lambda row: calc_run_diff(row['RS'], row['RA']), axis=1)
print(win_percs, '\n')

  Team League  Year   RS   RA   W    G  Playoffs   RD  RD_apply
0  SFG     NL  2012  718  649  94  162         1   69        69
1  SFG     NL  2011  570  578  86  162         0   -8        -8
2  SFG     NL  2010  697  583  92  162         1  114       114
3  SFG     NL  2009  657  611  88  162         0   46        46
4  SFG     NL  2008  640  759  72  162         0 -119      -119
0     69
1     -8
2    114
3     46
4   -119
dtype: int64 



## Second Most Efficient Method: `itertuples`

In [29]:
import numpy as np
def predict_win_perc(RS, RA):
    prediction = RS ** 2 / (RS ** 2 + RA ** 2)
    return np.round(prediction, 2)

win_perc_preds_loop = []

# Use a loop and .itertuples() to collect each row's predicted win percentage
for row in giants_df.itertuples():
    runs_scored = row.RS
    runs_allowed = row.RA
    win_perc_pred = predict_win_perc(runs_scored, runs_allowed)
    win_perc_preds_loop.append(win_perc_pred)

# Apply predict_win_perc to each row of the DataFrame
win_perc_preds_apply = giants_df.apply(lambda row: predict_win_perc(row['RS'], row['RA']), axis=1)
print(win_perc_preds_apply )

0    0.55
1    0.49
2    0.59
3    0.54
4    0.42
dtype: float64


## Using Broadcasting as the Most Efficient DataFrame Method

In [30]:

#* FASTEST SOLUTION
# Calculate the win percentage predictions using NumPy arrays
win_perc_preds_np = predict_win_perc(giants_df['RS'].values, giants_df['RA'].values)
giants_df['WP_preds'] = win_perc_preds_np
print(giants_df.head())

  Team League  Year   RS   RA   W    G  Playoffs   RD  RD_apply  WP_preds
0  SFG     NL  2012  718  649  94  162         1   69        69      0.55
1  SFG     NL  2011  570  578  86  162         0   -8        -8      0.49
2  SFG     NL  2010  697  583  92  162         1  114       114      0.59
3  SFG     NL  2009  657  611  88  162         0   46        46      0.54
4  SFG     NL  2008  640  759  72  162         0 -119      -119      0.42


Order of efficiency:

1. Broadcasting
2. `itertuples`
3. `apply`